In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

# Загрузка датасета

Загрузим датасет, настроим pandas на отображение всех колонок

In [2]:
dataset = pd.read_csv("dataset.csv", sep=';', decimal=',', index_col=0)

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

Первые записи:

In [3]:
dataset.head()

nplv                   DT  МАРКА ПРОФИЛЬ  t вып-обр  t обработка  \
0  г6837?3<87?  2070-06-21 00:14:00  Э76ХФ     Р65       29.0    45.366667   
1  г683863<886  2070-06-21 01:07:00  Э76ХФ     Р65       26.0    44.066667   
2  г683863<887  2070-06-21 02:00:00  Э76ХФ     Р65       24.0    43.350000   
3  г683863<888  2070-06-21 02:58:00  Э76ХФ     Р65       17.0    46.183333   
4  г683863<889  2070-06-21 03:50:00  Э76ХФ     Р65       20.0    48.500000   

   t под током  t продувка  ПСН гр.  чист расход C  чист расход Cr  \
0    24.400000   41.033333      NaN       0.456460        0.059572   
1    13.866667   44.066667      NaN       0.359285        0.083738   
2    17.950000   43.350000      NaN       0.331665        0.081490   
3    19.816667   46.183333      NaN       0.377945        0.133194   
4    17.033333   48.500000      NaN       0.389875        0.105094   

   чист расход Mn  чист расход Si  чист расход V  температура первая  \
0        0.117446        0.104762       0.040938              1557.0   
1        0.160923        0.110327       0.040083              1601.0   
2        0.132332        0.139860       0.041622              1593.0   
3        0.221605        0.165186       0.042050              1589.0   
4        0.169459        0.143024       0.040967              1597.0   

   температура последняя  Ar (интенс.)  N2 (интенс.)  эл. энергия (интенс.)  \
0                 1580.0     13.606742           NaN           12809.016393   
1                 1591.0      8.074721           NaN           12816.346154   
2                 1586.0     13.801968           NaN           12511.420613   
3                 1589.0     12.664959           NaN           12998.149706   
4                 1592.0     10.298350           NaN           12987.475538   

   произв жидкая сталь  произв количество обработок  \
0                115.5                          1.0   
1                111.6                          1.0   
2                115.8                          1.0   
3                116.3                          1.0   
4                115.0                          1.0   

   произв  количество плавок  произв количество плавок (цел)  расход газ Ar  \
0                        1.0                             1.0         18.611   
1                        1.0                             1.0         11.659   
2                        1.0                             1.0         19.871   
3                        1.0                             1.0         19.497   
4                        1.0                             1.0         16.649   

   расход газ  N2  расход C пров.  сыпуч известь РП  сыпуч кварцит  \
0             NaN             NaN               NaN            NaN   
1             NaN             NaN               NaN            NaN   
2             NaN             NaN               NaN            NaN   
3             NaN             NaN             0.077            NaN   
4             NaN             NaN             0.082            NaN   

   сыпуч кокс пыль УСТК  сыпуч  кокс. мелочь (сух.)  сыпуч  кокс. мелочь КМ1  \
0                 0.170                         NaN                    0.401   
1                 0.488                         NaN                      NaN   
2                 0.050                         NaN                    0.346   
3                 0.050                         NaN                    0.390   
4                 0.050                         NaN                    0.412   

   сыпуч  шпат плав.  ферспл CaC2  ферспл FeMo  ферспл FeSi-75  \
0              0.123         0.02          NaN             NaN   
1              0.094         0.02          NaN             NaN   
2              0.061         0.02          NaN             NaN   
3              0.059         0.02          NaN             NaN   
4              0.035         0.02          NaN             NaN   

   ферспл FeV азот.  ферспл  FeV-80  ферспл  Mn5Si65Al0.5  \
0               NaN        0.051939           

In [4]:
dataset.shape

(7041, 84)

Датасет содержит 7041 объект.
Объект содержит 84 признака.

Описание признаков:

In [5]:
dataset.describe(include='all')

nplv                   DT  МАРКА ПРОФИЛЬ    t вып-обр  \
count          7041                 7041   7041    7041  6444.000000   
unique         7041                 7041     18      18          NaN   
top     г683863=68=  2071-04-10 06:32:00  Э76ХФ     Р65          NaN   
freq              1                    1   4953    6427          NaN   
mean            NaN                  NaN    NaN     NaN    26.492241   
std             NaN                  NaN    NaN     NaN    14.842045   
min             NaN                  NaN    NaN     NaN    -1.000000   
25%             NaN                  NaN    NaN     NaN    14.000000   
50%             NaN                  NaN    NaN     NaN    25.000000   
75%             NaN                  NaN    NaN     NaN    37.000000   
max             NaN                  NaN    NaN     NaN    98.000000   

        t обработка  t под током   t продувка      ПСН гр.  чист расход C  \
count   6440.000000  6440.000000  6444.000000  5249.000000    6435.000000   
unique          NaN          NaN          NaN          NaN            NaN   
top             NaN          NaN          NaN          NaN            NaN   
freq            NaN          NaN          NaN          NaN            NaN   
mean      46.194099    21.010523    46.063051     3.004414       0.403088   
std        9.878727     4.758272     9.625492     0.738573       0.091197   
min        4.350000     2.816667     4.350000     1.000000       0.004400   
25%       42.350000    18.033333    42.216667     2.539699       0.345045   
50%       44.683333    20.500000    44.550000     3.000000       0.402250   
75%       47.550000    23.300000    47.516667     3.500875       0.460938   
max      294.083333    67.583333   251.000000     6.702208       0.909300   

        чист расход Cr  чист расход Mn  чист расход Si  чист расход V  \
count      5883.000000     6431.000000     6431.000000    6178.000000   
unique             NaN             NaN             NaN            NaN   
top                NaN             NaN             NaN            NaN   
freq               NaN             NaN             NaN            NaN   
mean          0.117692        0.139068        0.130561       0.051366   
std           0.031589        0.052807        0.041268       0.022126   
min           0.000572        0.000356        0.009790       0.002965   
25%           0.097838        0.103651        0.103468       0.039122   
50%           0.118770        0.138111        0.127332       0.040082   
75%           0.138435        0.172799        0.153948       0.042257   
max           0.675108        0.374175        0.478882       0.110047   

        температура первая  температура последняя  Ar (интенс.)  N2 (интенс.)  \
count          6436.000000            6436.000000   6426.000000   1389.000000   
unique                 NaN                    NaN           NaN           NaN   
top                    NaN                    NaN           NaN           NaN   
freq                   NaN                    NaN           NaN           NaN   
mean           1567.771753            1587.973897     11.593261     13.227743   
std              23.122019              14.714953      3.516874      4.058446   
min            1259.000000            1301.000000      2.456416      2.256585   
25%            1552.000000            1579.000000      9.167960     10.223178   
50%            1568.000000            1585.000000     11.039425     12.878459   
75%            1583.000000            1594.000000     13.474756     15.584599   
max            1673.000000            1691.000000     28.126298     27.997753   

        эл. энергия (интенс.)  произв жидкая сталь  \
count             6440.000000          6444.000000   
unique                    NaN                  NaN   
top                       NaN                  NaN   
freq                      NaN                  NaN   
mean             12989.038791           114.404252   
std                889.488799             2.357373 

In [6]:
dataset.columns

Index(['nplv', 'DT', 'МАРКА', 'ПРОФИЛЬ', 't вып-обр', 't обработка',
       't под током', 't продувка', 'ПСН гр.', 'чист расход C',
       'чист расход Cr', 'чист расход Mn', 'чист расход Si', 'чист расход V',
       'температура первая', 'температура последняя', 'Ar (интенс.)',
       'N2 (интенс.)', 'эл. энергия (интенс.)', 'произв жидкая сталь',
       'произв количество обработок', 'произв  количество плавок',
       'произв количество плавок (цел)', 'расход газ Ar', 'расход газ  N2',
       'расход C пров.', 'сыпуч известь РП', 'сыпуч кварцит',
       'сыпуч кокс пыль УСТК', 'сыпуч  кокс. мелочь (сух.)',
       'сыпуч  кокс. мелочь КМ1', 'сыпуч  шпат плав.', 'ферспл CaC2',
       'ферспл FeMo', 'ферспл FeSi-75', 'ферспл FeV азот.', 'ферспл  FeV-80',
       'ферспл  Mn5Si65Al0.5', 'ферспл  Ni H1 пласт.', 'ферспл SiMn18',
       'ферспл  ферванит', 'ферспл  фх850А', 'эл. энергия',
       'химсталь первый Al_1', 'химсталь первый C_1', 'химсталь первый Cr_1',
       'химсталь первый 

## Целевые признаки

In [12]:
# целевые признаки (принадлежат к вещественному типу)
target_features = ['химшлак последний Al2O3', 'химшлак последний CaO',
       'химшлак последний FeO', 'химшлак последний MgO',
       'химшлак последний MnO', 'химшлак последний R',
       'химшлак последний SiO2']

# Количество пропусков и типы данных

In [7]:
def make_description(df):
    description = pd.concat([df.isna().sum(), df.dtypes, df.nunique()], axis=1)
    description.rename(columns={0: "num of NaN", 1: "dtypes", 2: "nunique"}, inplace=True)
    return description

In [8]:
make_description(dataset).sort_values('num of NaN', ascending=False)

num of NaN   dtypes  nunique
ферспл FeSi-75                        7036  float64        2
сыпуч  кокс. мелочь (сух.)            7034  float64        7
сыпуч кварцит                         7028  float64        1
ферспл FeMo                           7008  float64       27
ферспл  Ni H1 пласт.                  6959  float64       80
ферспл FeV азот.                      6658  float64       79
расход C пров.                        6317  float64      589
ферспл  ферванит                      6024  float64      194
расход газ  N2                        5652  float64     1341
N2 (интенс.)                          5652  float64     1389
химшлак последний SiO2                2299  float64      116
химшлак последний Al2O3               2299  float64       70
химшлак первый SiO2_1                 2299  float64      149
химшлак первый Al2O3_1                2299  float64       73
сыпуч известь РП                      2271  float64      441
ПСН гр.                               1792  float64     4296
чист расход Cr                        1158  float64     3115
ферспл  фх850А                        1158  float64      370
сыпуч  шпат плав.                     1131  float64      177
чист расход V                          863  float64     4645
ферспл  FeV-80                         862  float64      650
химсталь последний Ca                  823  float64       32
ферспл CaC2                            797  float64        6
сыпуч  кокс. мелочь КМ1                734  float64      652
ферспл SiMn18                          643  float64      497
ферспл  Mn5Si65Al0.5                   641  float64      368
сыпуч кокс пыль УСТК                   619  float64      306
Ar (интенс.)                           615  float64     6423
расход газ Ar                          615  float64     5548
чист расход Mn                         610  float64     6320
чист расход Si                         610  float64     6265
чист расход C                          606  float64     6243
температура первая                     605  float64      154
температура последняя                  605  float64      109
эл. энергия                            601  float64     3132
t обработка                            601  float64      816
эл. энергия (интенс.)                  601  float64     6362
t под током                            601  float64     1155
химсталь последний N                   599  float64       17
химсталь первый Mo_1                   598  float64       24
химшлак последний MgO                  598  float64      192
химшлак первый MgO_1                   598  float64      236
химсталь первый N_1                    598  float64       18
t вып-обр                              597  float64       82
химсталь последний Ni                  597  float64      112
химсталь последний P                   597  float64       19
химсталь последний S                   597  float64       27
химсталь последний Si                  597  float64      256
химсталь последний Ti                  597  float64        5
химсталь последний V                   597  float64       54
произв количество плавок (цел)         597  float64        1
химшлак первый CaO_1                   597  float64      298
химсталь последний Mn                  597  float64      261
химшлак первый FeO_1                   597  float64       54
химшлак первый MnO_1                   597  float64      152
химшлак первый R_1                     597  float64       32
химшлак последний CaO                  597  float64      229
химшлак последний FeO                  597  float64       16
химшлак последний MnO                  597  float64       61
химшлак последний R                    597  float64       16
химсталь последний Mo                  597  float64       24
химсталь последний Cr                  597  float64      206
химсталь последний Cu                  597  float64      100
химсталь первый Mn_1                   597  float64      404
произв количество обработок            597  float64        3
произв

## Пропуски по столбцам
Как видно, некоторные признаки имеют значительное количество пропусков (более 33%)
Также важно отметить, что некоторые из признаков содержат всего 1 уникальное значение помимо NaN.

In [10]:
def list_features_with_single_value(df):
    features_list = []
    for feature in df.columns:
        if df[feature].nunique() == 1:
            features_list.append(feature)
    return features_list


def list_features_with_lots_of_nan(df):
    features_list = []
    for feature in df.columns:
        if feature in target_features: # не рассматриваем целевые признаки
            continue
        num_of_nan = dataset[feature].isna().sum()
        if num_of_nan > dataset.shape[0] / 3:
            features_list.append(feature)
    return features_list


In [13]:
features_with_lots_of_nan = list_features_with_lots_of_nan(dataset)

print("Столбцы с большим количеством пропусков:")
print(features_with_lots_of_nan)


features_with_single_value = list_features_with_single_value(dataset)

print("Признаки с единственным уникальным значением:")
print(features_with_single_value)

Столбцы с большим количеством пропусков:
['N2 (интенс.)', 'расход газ  N2', 'расход C пров.', 'сыпуч кварцит', 'сыпуч  кокс. мелочь (сух.)', 'ферспл FeMo', 'ферспл FeSi-75', 'ферспл FeV азот.', 'ферспл  Ni H1 пласт.', 'ферспл  ферванит']
Признаки с единственным уникальным значением:
['произв  количество плавок', 'произв количество плавок (цел)', 'сыпуч кварцит']


## Пропуски по строкам
Узнаем количество строк, в которых отсутствует более 33% признаков:

In [14]:
def list_rows_with_lots_of_nan(df):
    rows_list = []
    numbers_of_nan = df.isna().sum(axis=1)
    for index, count in enumerate(numbers_of_nan):
        if count > df.shape[1] / 3:
            rows_list.append(index)
    return rows_list

In [15]:
rows_with_lots_of_nan = list_rows_with_lots_of_nan(dataset)
print("Строк с большим количеством пропусков: {}".format(len(rows_with_lots_of_nan)))

Строк с большим количеством пропусков: 599


## Категориальные признаки

Категориальными типами являются:
- **nplv** - номер плавки, не имеет смысла
- **DT** - дата и время плавки
- **МАРКА** - марка стали (?)
- **ПРОФИЛЬ** - профиль стали (?)

In [16]:
categorical_features = ["nplv", "DT", "МАРКА", "ПРОФИЛЬ"]

dataset[categorical_features].head()

nplv                   DT  МАРКА ПРОФИЛЬ
0  г6837?3<87?  2070-06-21 00:14:00  Э76ХФ     Р65
1  г683863<886  2070-06-21 01:07:00  Э76ХФ     Р65
2  г683863<887  2070-06-21 02:00:00  Э76ХФ     Р65
3  г683863<888  2070-06-21 02:58:00  Э76ХФ     Р65
4  г683863<889  2070-06-21 03:50:00  Э76ХФ     Р65

In [17]:
print("Количество всех пропущенных категориальных значений: {}".format(dataset[categorical_features].isna().sum().sum()))

Количество всех пропущенных категориальных значений: 0


#### МАРКА

In [18]:
print("Количество уникальных значений: {}".format(dataset["МАРКА"].nunique()))
print("\nЧастота встречаемости каждого из значений:")
print(dataset["МАРКА"].value_counts())

Количество уникальных значений: 18

Частота встречаемости каждого из значений:
Э76ХФ     4953
Э90ХАФ    1384
Э73Ф       227
Э76Ф        86
Ш2.2        59
Э76         56
900А        54
R350HT      54
Э76ХАФ      50
R260        35
G2HH        34
Ш2Л         31
НХ           7
09Г2С        4
С345         3
С255         2
Ст3сп        1
С355         1
Name: МАРКА, dtype: int64


#### ПРОФИЛЬ

In [19]:
print("Количество уникальных значений: {}".format(dataset["ПРОФИЛЬ"].nunique()))
print("\nЧастота встречаемости каждого из значений:")
print(dataset["ПРОФИЛЬ"].value_counts())

Количество уникальных значений: 18

Частота встречаемости каждого из значений:
Р65                                 6427
ОР65                                 221
60E1                                  95
Шар заг d шара 80-100 мм 2,3 гр.      59
Рельсы трамвайные                     56
UIC54                                 54
Р50                                   44
Шар заг d шара 80-100 мм 4 гр.        17
136 RE                                17
Шары мелющие d 80-100 мм              14
54E1                                  10
НЕТ                                    7
ОР50                                   6
Швеллер 40У                            6
Швеллер 40У (12 категория)             4
Шар заг d шара 100 мм 2 гр.            2
60E2                                   1
НЛЗ 300х365                            1
Name: ПРОФИЛЬ, dtype: int64


## Вещественные признаки

Все остальные 80 признаков имеют вещественный тип данных

In [20]:
numerical_features = ['t вып-обр', 't обработка',
       't под током', 't продувка', 'ПСН гр.', 'чист расход C',
       'чист расход Cr', 'чист расход Mn', 'чист расход Si', 'чист расход V',
       'температура первая', 'температура последняя', 'Ar (интенс.)',
       'N2 (интенс.)', 'эл. энергия (интенс.)', 'произв жидкая сталь',
       'произв количество обработок', 'произв  количество плавок',
       'произв количество плавок (цел)', 'расход газ Ar', 'расход газ  N2',
       'расход C пров.', 'сыпуч известь РП', 'сыпуч кварцит',
       'сыпуч кокс пыль УСТК', 'сыпуч  кокс. мелочь (сух.)',
       'сыпуч  кокс. мелочь КМ1', 'сыпуч  шпат плав.', 'ферспл CaC2',
       'ферспл FeMo', 'ферспл FeSi-75', 'ферспл FeV азот.', 'ферспл  FeV-80',
       'ферспл  Mn5Si65Al0.5', 'ферспл  Ni H1 пласт.', 'ферспл SiMn18',
       'ферспл  ферванит', 'ферспл  фх850А', 'эл. энергия',
       'химсталь первый Al_1', 'химсталь первый C_1', 'химсталь первый Cr_1',
       'химсталь первый Cu_1', 'химсталь первый Mn_1', 'химсталь первый Mo_1',
       'химсталь первый N_1', 'химсталь первый Ni_1', 'химсталь первый P_1',
       'химсталь первый S_1', 'химсталь первый Si_1', 'химсталь первый Ti_1',
       'химсталь первый V_1', 'химсталь последний Al', 'химсталь последний C',
       'химсталь последний Ca', 'химсталь последний Cr',
       'химсталь последний Cu', 'химсталь последний Mn',
       'химсталь последний Mo', 'химсталь последний N',
       'химсталь последний Ni', 'химсталь последний P', 'химсталь последний S',
       'химсталь последний Si', 'химсталь последний Ti',
       'химсталь последний V', 'химшлак первый Al2O3_1',
       'химшлак первый CaO_1', 'химшлак первый FeO_1', 'химшлак первый MgO_1',
       'химшлак первый MnO_1', 'химшлак первый R_1', 'химшлак первый SiO2_1',
       'химшлак последний Al2O3', 'химшлак последний CaO',
       'химшлак последний FeO', 'химшлак последний MgO',
       'химшлак последний MnO', 'химшлак последний R',
       'химшлак последний SiO2']

In [21]:
dataset[numerical_features].shape

(7041, 80)

### Количество NaN значений у каждого вещественного признака

In [22]:
dataset[numerical_features].isna().sum()

t вып-обр                          597
t обработка                        601
t под током                        601
t продувка                         597
ПСН гр.                           1792
чист расход C                      606
чист расход Cr                    1158
чист расход Mn                     610
чист расход Si                     610
чист расход V                      863
температура первая                 605
температура последняя              605
Ar (интенс.)                       615
N2 (интенс.)                      5652
эл. энергия (интенс.)              601
произв жидкая сталь                597
произв количество обработок        597
произв  количество плавок          597
произв количество плавок (цел)     597
расход газ Ar                      615
расход газ  N2                    5652
расход C пров.                    6317
сыпуч известь РП                  2271
сыпуч кварцит                     7028
сыпуч кокс пыль УСТК               619
сыпуч  кокс. мелочь (сух.

# Распределение целевых признаков

Изначальное, до чистки:

In [23]:
for feature in target_features:
    #sns.displot(dataset[feature])
    print(feature)
    print("Min: {}, max: {}".format(dataset[feature].min(), dataset[feature].max()))
    print("Mean: {}, median: {}, mode: {}, std: {}".format(dataset[feature].mean(),
                                                           dataset[feature].median(),
                                                           dataset[feature].mode()[0],
                                                           dataset[feature].std()))
    print("Number of NaN: {}".format(dataset[feature].isna().sum()))
    print()

химшлак последний Al2O3
Min: 2.4, max: 14.2
Mean: 4.59791227330243, median: 4.5, mode: 4.4, std: 0.9601234421428474
Number of NaN: 2299

химшлак последний CaO
Min: 35.9, max: 67.5
Mean: 56.38299193047791, median: 56.8, mode: 57.6, std: 3.7800368454266327
Number of NaN: 597

химшлак последний FeO
Min: 0.1, max: 4.2
Mean: 0.5857231533209227, median: 0.6, mode: 0.5, std: 0.21691848413805986
Number of NaN: 597

химшлак последний MgO
Min: 0.3, max: 32.4
Mean: 9.061989756324689, median: 8.9, mode: 8.7, std: 3.163435967039685
Number of NaN: 598

химшлак последний MnO
Min: 0.04, max: 1.45
Mean: 0.12379422718808451, median: 0.11, mode: 0.08, std: 0.06458709833264452
Number of NaN: 597

химшлак последний R
Min: 1.6, max: 3.1
Mean: 2.3614245810056675, median: 2.3, mode: 2.2, std: 0.22417154717293655
Number of NaN: 597

химшлак последний SiO2
Min: 16.4, max: 28.7
Mean: 23.971699704765893, median: 24.4, mode: 26.0, std: 2.1877678629082578
Number of NaN: 2299



После чистки:

In [24]:
features_to_drop = list(set(features_with_lots_of_nan + features_with_single_value))
preprocessed_dataset = dataset.drop(features_to_drop, axis=1)
preprocessed_dataset = preprocessed_dataset.drop(rows_with_lots_of_nan, axis=0)

print("Размер датасета после первичной чистки: {}\n".format(preprocessed_dataset.shape))

for feature in target_features:
    #sns.displot(dataset[feature])
    print(feature)
    print("Min: {}, max: {}".format(preprocessed_dataset[feature].min(), preprocessed_dataset[feature].max()))
    print("Mean: {}, median: {}, mode: {}, std: {}".format(preprocessed_dataset[feature].mean(),
                                                           preprocessed_dataset[feature].median(),
                                                           preprocessed_dataset[feature].mode()[0],
                                                           preprocessed_dataset[feature].std()))
    print("Number of NaN: {}".format(preprocessed_dataset[feature].isna().sum()))
    print()

Размер датасета после первичной чистки: (6442, 72)

химшлак последний Al2O3
Min: 2.4, max: 14.2
Mean: 4.597637629192178, median: 4.5, mode: 4.4, std: 0.9600384070032368
Number of NaN: 1701

химшлак последний CaO
Min: 35.9, max: 67.5
Mean: 56.38283141881399, median: 56.8, mode: 57.6, std: 3.780532200104373
Number of NaN: 0

химшлак последний FeO
Min: 0.1, max: 4.2
Mean: 0.585749767153062, median: 0.6, mode: 0.5, std: 0.2169468988853767
Number of NaN: 0

химшлак последний MgO
Min: 0.3, max: 32.4
Mean: 9.061263778916313, median: 8.9, mode: 8.7, std: 3.1635880965844794
Number of NaN: 1

химшлак последний MnO
Min: 0.04, max: 1.45
Mean: 0.12378453896305748, median: 0.11, mode: 0.08, std: 0.06459472372082203
Number of NaN: 0

химшлак последний R
Min: 1.6, max: 3.1
Mean: 2.3614436510401307, median: 2.3, mode: 2.2, std: 0.2241968099724476
Number of NaN: 0

химшлак последний SiO2
Min: 16.4, max: 28.7
Mean: 23.971609365112812, median: 24.4, mode: 26.0, std: 2.1879897822052676
Number of NaN: 1701
